### PR TITLE
DONE: add parent/child roles including debug pages

### PR DESCRIPTION
**Pull request by reinout**
@remcogerlich: WIP, I want to add a couple of test cases. But the actual code is there and ready for some feedback :-)

But... it seems to be working. I first copied the testcase from your branch and adjusted it to my code. The actual test that's done is the same.

The relation is a regular m2m from role directly to role instead of an intermediary model. I called it parent/child as I kept disliking other naming options and kept mixing up the names.

The `all_organisation_roles()` doesn't need any python processing: it is The Mother Of All Queries. Heavily documented by using Q objects. In the end, it just fires off one query. I still have to check it for complete correctness, though. And I have to get the djangodebugtoolbar running in SSO so I can check the query.

### CODE REVIEW

**Comment by reinout** [comment]
@remcogerlich: ik ben er klaar mee, volgens mij. Kom anders vrijdagochtend een keertje buurten, dan kan ik het laten zien.

File: lizard_auth_server/tests/test_models.py [hunk 1]
```diff
@@ -99,6 +100,71 @@ def test_only_the_given_portal_is_returned(self):
...
+
+        # See, he doesn't have any roles on portal2:
+        self.assertEquals(
+            len(list(profile.all_organisation_roles(portal2))), 0)
+
+        # However, when the second role an inheriting role of the first (which
+        # then becomes the second role's base role):
+        role1.inheriting_roles.add(role2)
+
+        profile = models.UserProfile.objects.fetch_for_user(user)
```
Comment (reinout) [inline lizard_auth_server/tests/test_models.py]: Kan misschien weg

File: lizard_auth_server/models.py [hunk 2]
```diff
@@ -313,19 +314,62 @@ def has_access(self, portal):
...
+        relevant_role_and_direct_access = models.Q(
+            id__in=organisation_roles_i_can_access,
+            role__in=relevant_roles_tied_to_the_portal)
+
+        # The elaborate case is that a role must of course be a relevant
+        # role. Then that same role must have a base role with an organisation
+        # role that I can access. That same organisation role must also have
+        # the same organisation as the organisation role I'm looking
+        # from. Django ensures those "the same" items are really the
+        # same. This needs more tests, though.
```
Comment (reinout) [inline lizard_auth_server/models.py]: Comment fixen

File: lizard_auth_server/models.py [hunk 3]
```diff
@@ -313,19 +314,62 @@ def has_access(self, portal):
...
+            organisation_roles_directly = OrganisationRole.objects.filter(
+                tied_to_my_user_profile)
+            organisation_roles_via_organisation = OrganisationRole.objects.filter(
+                tied_to_my_organisation_for_all_users).distinct()
+            direct_results = OrganisationRole.objects.filter(
+                relevant_role_and_direct_access).distinct()
+            indirect_results = OrganisationRole.objects.filter(
+                relevant_role_and_indirect_access_with_matching_org).distinct()
+
+            return locals()
```
Comment (remcogerlich) [inline lizard_auth_server/models.py]: En toch vind ik locals() niet mooi. Dan hangt code ergens anders dus af van die variabelenamen hier. Misschien een expliciete dict returnen?

**Comment by remcogerlich** [comment]
Er staan nog een paar kleine commentaar items open.

En ik denk dat de verbinding van users aan portals weg kan, dat is gewoon dubbel (doen we al uitgebreid via rollen), en ik denk dat niemand echt beseft dat dat er ook nog is. Vragen aan Lex of hij het erg vindt als het weg gaat?

File: lizard_auth_server/models.py (hunk 3 above)
Comment (reinout) [inline lizard_auth_server/models.py]: Je hebt daar wel gelijk in. Ik ga even een dict bakken.

Comment (reinout) [inline lizard_auth_server/models.py]: Dict gebakken :-)

**Comment by reinout** [comment]
Oei, het betekent wel dat die hele org-role query uitgevoerd moet worden. Maar inderdaad, vraag Lex maar.

Aan de andere kant... kan nog niet weg. Veel sites werken niet met rollen en org rollen. Bij efcis hebben we alleen gebruikers aan een portaal gekoppeld en dat was dat. Al die rollen-administratie is daar niet nodig. Idem trs. Idem vast nog een bak sites.

**Comment by reinout** [comment]
Ik heb de debug toolbar erin hangen, dus maandag fix ik nog wat qua select_related :-)

Ik heb er trouwens een duidelijke feature aan toegevoegd: https://github.com/lizardsystem/lizard-auth-server/commit/4ba982ec5b19a604b2943348d5047b74fd8336af . Een automatische select_related omdat die me heeeeeel vaak nodig leek, gezien de debug toolbar...

**Comment by reinout** [comment]
Voor de lol bij deze de dikke vette query die afgevuurd wordt :-)

SELECT DISTINCT "lizard_auth_server_organisationrole"."id", "lizard_auth_server_organisationrole"."organisation_id", "lizard_auth_server_organisationrole"."role_id", "lizard_auth_server_organisationrole"."for_all_users", "lizard_auth_server_organisation"."id", "lizard_auth_server_organisation"."name", "lizard_auth_server_organisation"."unique_id", "lizard_auth_server_role"."id", "lizard_auth_server_role"."portal_id", "lizard_auth_server_role"."unique_id", "lizard_auth_server_role"."code", "lizard_auth_server_role"."name", "lizard_auth_server_role"."external_description", "lizard_auth_server_role"."internal_description", "lizard_auth_server_portal"."id", "lizard_auth_server_portal"."name", "lizard_auth_server_portal"."sso_secret", "lizard_auth_server_portal"."sso_key", "lizard_auth_server_portal"."allowed_domain", "lizard_auth_server_portal"."redirect_url", "lizard_auth_server_portal"."visit_url" FROM "lizard_auth_server_organisationrole" INNER JOIN "lizard_auth_server_role" ON ( "lizard_auth_server_organisationrole"."role_id" = "lizard_auth_server_role"."id" ) INNER JOIN "lizard_auth_server_organisation" ON ( "lizard_auth_server_organisationrole"."organisation_id" = "lizard_auth_server_organisation"."id" ) LEFT OUTER JOIN "lizard_auth_server_role_inheriting_roles" ON ( "lizard_auth_server_role"."id" = "lizard_auth_server_role_inheriting_roles"."to_role_id" ) LEFT OUTER JOIN "lizard_auth_server_role" T5 ON ( "lizard_auth_server_role_inheriting_roles"."from_role_id" = T5."id" ) LEFT OUTER JOIN "lizard_auth_server_organisationrole" T6 ON ( T5."id" = T6."role_id" ) INNER JOIN "lizard_auth_server_portal" ON ( "lizard_auth_server_role"."portal_id" = "lizard_auth_server_portal"."id" ) WHERE (("lizard_auth_server_organisationrole"."role_id" IN (SELECT "lizard_auth_server_role"."id" FROM "lizard_auth_server_role" WHERE "lizard_auth_server_role"."portal_id" = 52 ) AND "lizard_auth_server_organisationrole"."id" IN (SELECT U0."id" FROM "lizard_auth_server_organisationrole" U0 INNER JOIN "lizard_auth_server_organisation" U1 ON ( U0."organisation_id" = U1."id" ) LEFT OUTER JOIN "lizard_auth_server_userprofile_organisations" U2 ON ( U1."id" = U2."organisation_id" ) LEFT OUTER JOIN "lizard_auth_server_userprofile_roles" U4 ON ( U0."id" = U4."organisationrole_id" ) WHERE ((U2."userprofile_id" = 14 AND U0."for_all_users" = true ) OR U4."userprofile_id" = 14 ))) OR (T6."organisation_id" = "lizard_auth_server_organisationrole"."organisation_id" AND "lizard_auth_server_organisationrole"."role_id" IN (SELECT "lizard_auth_server_role"."id" FROM "lizard_auth_server_role" WHERE "lizard_auth_server_role"."portal_id" = 52 ) AND T6."id" IN (SELECT "lizard_auth_server_organisationrole"."id" FROM "lizard_auth_server_organisationrole" INNER JOIN "lizard_auth_server_organisation" ON ( "lizard_auth_server_organisationrole"."organisation_id" = "lizard_auth_server_organisation"."id" ) LEFT OUTER JOIN "lizard_auth_server_userprofile_organisations" ON ( "lizard_auth_server_organisation"."id" = "lizard_auth_server_userprofile_organisations"."organisation_id" ) LEFT OUTER JOIN "lizard_auth_server_userprofile_roles" ON ( "lizard_auth_server_organisationrole"."id" = "lizard_auth_server_userprofile_roles"."organisationrole_id" ) WHERE (("lizard_auth_server_userprofile_organisations"."userprofile_id" = 14 AND "lizard_auth_server_organisationrole"."for_all_users" = true ) OR "lizard_auth_server_userprofile_roles"."userprofile_id" = 14 ))))

**Pull request merged**